src/widget.rs

```diff
@@ -2,35 +2,97 @@
 // See the COPYRIGHT file at the top-level directory of this distribution.
 // Licensed under the MIT license, see the LICENSE file or <http://opensource.org/licenses/MIT>
 
+use std::boxed::Box as Box_;
 use std::mem::transmute;
+use std::ptr;
 
 use glib::object::{Downcast, IsA};
 use glib::signal::connect;
 use glib::translate::*;
 use glib_ffi::gboolean;
-use gdk::Event;
+use gdk::{DragAction, DragContext, Event, ModifierType};
 use gdk_ffi;
+use glib_ffi;
+use libc;
 use pango;
 use ffi;
 
 use {
+    DestDefaults,
     Inhibit,
     Object,
     Rectangle,
+    SelectionData,
+    TargetEntry,
     Widget,
 };
 
 pub trait WidgetExtManual {
+    fn drag_dest_set(&self, flags: DestDefaults, targets: &[TargetEntry], actions: DragAction);
+
+    fn drag_source_set(&self, start_button_mask: ModifierType, targets: &[TargetEntry], actions: DragAction);
+
     fn intersect(&self, area: &Rectangle, intersection: Option<&mut Rectangle>) -> bool;
 
     fn override_font(&self, font: &pango::FontDescription);
 
+    fn connect_drag_data_get<F: Fn(&Self, &DragContext, &mut SelectionData, u32, u32) + 'static>(&self, f: F) -> u64;
+
     fn connect_map_event<F: Fn(&Self, &Event) -> Inhibit + 'static>(&self, f: F) -> u64;
 
     fn connect_unmap_event<F: Fn(&Self, &Event) -> Inhibit + 'static>(&self, f: F) -> u64;
 }
 
 impl<O: IsA<Widget> + IsA<Object>> WidgetExtManual for O {
+
+    fn drag_dest_set(&self, flags: DestDefaults, targets: &[TargetEntry], actions: DragAction) {
+        let stashes: Vec<_> = targets.iter().map(|e| e.to_glib_none()).collect();
+        let mut t = Vec::with_capacity(stashes.len());
+        for stash in stashes {
```

Comment:
Found problem by accident: without `&stashes` it moved to for and dropped afterward.